README.md
# Lighthouse runner for python

## About This Repository
This repository is based on the original work of [Adam Cupiał](https://github.com/adamcupial) and was forked from his [lighthouse-python](https://github.com/adamcupial/lighthouse-python) repository. While the original repository had not been updated in five years, I found it to be a valuable resource and decided to update it (I have made some changes, fixed some bugs) and turn it into a package for PyPI. I want to acknowledge the original author for creating the foundation of this project and making it available to the community.


## Description
This library is a simple wrapper around lighthouse-cli runner that runs the audit and parses a result in friendly manner.

## Installation
if lighthouse is not installed:
```bash
npm install -g lighthouse
```

```bash
pip install lighthouse-python-plus
```

## Usage

```python
from lighthouse import LighthouseRunner

TIMINGS = [
    'speed-index'
]

report = LighthouseRunner('https://github.com/adamcupial', form_factor='desktop', quiet=False, timings=TIMINGS).report
assert report.score['performance'] > 0.5
print(report.audits(0.5)['performance'].failed)
```

report has 3 properties:

- score: returns dict where keys are categories and values are scores (0 to 1)
- timings: returns dict where keys are timings and values are timedelta objects
- audits: dict where keys are categories and values are objects with passed and failed lists attached

To see all the usage examples of this library, please check out the [demo](demo) folder.

## Dependencies
 - python 3.6+
 - node package lighthouse>=10.3.0 installed

## Changes
[You can find all changes in CHANGELOG!](CHANGELOG.md)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.2.0] - 2023-06-29
## changed
- All modules receive timings as input
- The raw output can be saved in a preferred format, either json, html, or both
- The demo has been updated

## [1.1.0] - 2023-06-29
## changed
- Old syntax fixed
- Made compatible to run in docker
- Demo Added

## Adam Cupal
## [1.0.2] - 2019-02-22
### Changed
- added option to runner to pass custom parameters

## [1.0.1] - 2019-02-14
### Changed
- When error occurs during run the process should be stopped (issue #1)
- Added changelog

## [1.0.0] - 2019-02-05
### Added
- Base runner


tests/test_batch_runner.py
from lighthouse import BatchRunner


urls = ["https://amazon.com","https://alibaba.com"]
form_factors = ["mobile","desktop"]

# settings are as the same as the pagespeed.web.dev
additional_settings = [
"--throttling.cpuSlowdownMultiplier=4",
"--throttling.connectionType=4g",
"--throttling.downloadThroughputKbps=1638.4",
"--throttling.rttMs=150",
"--screenEmulation.disabled=false",
"--screenEmulation.width=412",
"--screenEmulation.height=823",
"--screenEmulation.deviceScaleFactor=1.75",
"--axe.enable=true",
"--axe.version=4.7.0",
]

batch_runner = BatchRunner(urls,form_factors,quiet=False,additional_settings=additional_settings,repeats=2)

print(batch_runner.reports)


lighthouse/batch.py
# Python Standard Library
from itertools import product

# Third party modules
from tqdm import tqdm

# Local imports
from .runner import LighthouseRepeatRunner


class BatchRunner(object):
    def __init__(self, urls, form_factors, quiet=True,
                 additional_settings=None, repeats=3):

        if not isinstance(form_factors, (list, )):
            form_factors = list(form_factors)
        if not isinstance(urls, (list, )):
            urls = list(urls)

        all_combinations = tqdm(list(product(urls, form_factors)),
                                desc='Running urls')

        self.reports = []

        for url, factor in all_combinations:
            all_combinations.set_description('{0} | {1}'.format(url, factor))
            report = LighthouseRepeatRunner(url, factor, quiet,
                                            additional_settings,
                                            repeats).report
            self.reports.append((url, factor, report))



lighthouse/report.py
# Standard library modules
from collections import namedtuple
from datetime import timedelta
from typing import Dict, List


AuditResult = namedtuple('AuditResult', ['id','title', 'score', 'display','numval','numunit'])
AuditCategoryResult = namedtuple('AuditCategoryResult', ['passed', 'failed'])

class LighthouseReport:
    """
    Represents a Lighthouse report.

    Provides a more user-friendly interface to the Lighthouse JSON report.
    """

    def __init__(self, data: Dict, timings: List[str] = []):
        """
        Args:
            data (dict): JSON loaded Lighthouse report
            
            timings (list(str), optional): List of timings to gather
                from Lighthouse report. Defaults to an empty list.
        """

        self.__data = data
        self.__timings = timings

    @property
    def score(self) -> Dict[str, float]:
        """
        Dictionary of Lighthouse category names and their corresponding scores (0 to 1).
        """

        return {
            k: v.get('score', 0)
            for k, v
            in self.__data['categories'].items()
        }

    @property
    def timings(self) -> Dict[str, timedelta]:
        """
        Dictionary of Lighthouse timing names and their corresponding times as timedeltas.
        """

        return {
            k: timedelta(milliseconds=v.get('numericValue')) 
            for k, v
            in self.__data['audits'].items()
            if k in self.__timings
        }

    def audits(self, score_thresh: float = 1) -> Dict[str, AuditCategoryResult]:
        """
        Returns a dictionary of Lighthouse audits, grouped by category, with passed/failed
        keys and lists of passed/failed audits attached.

        Args:
            score_thresh (float, optional): The minimum score required for an audit to be considered
                passed. Defaults to 1 (all audits must pass).
        """

        res = {}
        for category, data in self.__data['categories'].items():
            all_audit_refs = [
                x.get('id')
                for x in data['auditRefs']
            ]
            all_audits = {k: self.__data['audits'][k] for k in all_audit_refs}
            sdm_to_reject = ['manual', 'notApplicable', 'informative']
            passed_audits = [
                AuditResult(**{
                    'id': v['id'],
                    'title': v['title'],
                    'score': v['score'],
                    'display': v.get('displayValue'),
                    'numval': v.get('numericValue'),
                    'numunit': v.get('numericUnit'),
                })
                for k, v in all_audits.items()
                if v.get('score',0) is not None and 
                v.get('score', 0) >= score_thresh and
                v.get('scoreDisplayMode') not in sdm_to_reject
            ]

            failed_audits = [
                AuditResult(**{
                    'id': v['id'],
                    'title': v['title'],
                    'score': v['score'],
                    'display': v.get('displayValue'),
                    'numval': v.get('numericValue'),
                    'numunit': v.get('numericUnit'),
                })
                for k, v in all_audits.items()
                if (v.get('score',0) is None or v.get('score', 0) < score_thresh) and
                v.get('scoreDisplayMode') not in sdm_to_reject
            ]

            res[category] = AuditCategoryResult(passed_audits, failed_audits)
        return res


lighthouse/runner.py
# Standard library modules
from collections import namedtuple
from datetime import datetime
from functools import reduce
from itertools import product
import json
import os
import shutil
import subprocess
from typing import List, Union, Dict, Any
from uuid import uuid4

# Third-party modules
from tqdm import tqdm

# Local modules
from .report import LighthouseReport


BASE_TIMINGS = [
    'first-contentful-paint',
    'speed-index',
    'interactive',
    'first-meaningful-paint',
    'first-cpu-idle',
    'estimated-input-latency',
    'time-to-first-byte',
]

class LighthouseRunner:
    """
    Lightweight runner, wraps around lighthouse-cli and parses the result

    Attributes:
        report (LighthouseReport or None): A LighthouseReport object containing the report data, or None if the report 
            could not be loaded.
    """

    def __init__(self, url:str, form_factor:str='mobile', quiet:bool=True,
                 additional_settings:List[str]=[],timings:List[str]=BASE_TIMINGS,
                 output_type:str='none',output_dir:str='./') -> None :
        """ 
        Initialize a new LighthousePerformanceTest object.
        
        Args:
            url (str): The URL to test.
            
            form_factor (str, optional): The form factor to use for the test. Either 'mobile' or 'desktop'.
                Defaults to 'mobile'.
                
            quiet (bool, optional): Whether to suppress output to stdout. Defaults to True.
            
            additional_settings (list, optional): A list of additional parameters to use for the test.
                Defaults to an empty list.
                
            timings (list, optional): A list of performance metrics to measure during the test.
                Defaults to BASE_TIMINGS, which includes several common metrics.
                
            output_type (str, optional): The type of output to generate. Possible values are 'none','both',
                'json', and 'html'. Defaults to 'none'.
                
            output_dir (str, optional): The directory where output files should be written.
                Defaults to the current directory.
        
        Raises:
            AssertionError: If form_factor or output_type are not valid values.
            
        """

        # Check inputs
        assert form_factor in ['mobile', 'desktop'], f"Invalid form_factor: {form_factor}. Allowed values are 'mobile' or 'desktop'."
        assert output_type in ['json','html','both','none'], f"Invalid output_type: {output_type}. Allowed values are 'json', 'html', 'both', or 'none'."
        
        # Generate a unique report name based on the current date and time, plus a random UUID for report path
        self.report_name = str(datetime.now()).replace(" ","_") +"_"+ uuid4().hex
        self.report_path = os.path.join("/tmp",self.report_name)
        
        # Set the timings, output directory, and output type for the test
        self.timings = timings
        self.output_dir = output_dir
        self.output_type = output_type.lower().strip()
        
        # Run the performance test with the given settings and get the report data
        self._run(url, form_factor, quiet, additional_settings)
        self.report = self._get_report()
        
        # Clean up any temporary files
        self._clean()

    def _run(self, url:str, form_factor:str='mobile', quiet:bool=True,
                 additional_settings:List[str]=[]) -> None:
        """
        Run a performance test using the Lighthouse CLI.
        
        Args:
            url (str): The URL to test.
            
            form_factor (str, optional): The form factor to use for the test. Either 'mobile' or 'desktop'.
                Defaults to 'mobile'.
                
            quiet (bool, optional): Whether to suppress output to stdout. Defaults to True.
            
            additional_settings (list, optional): A list of additional parameters to use for the test.
                Defaults to an empty list.

        Raises:
            RuntimeError: If there is an error running the performance test using the Lighthouse CLI, \
            such as a missing or invalid URL, an invalid form factor, a problem with the output path, or an issue with the Lighthouse CLI itself.
            
        """

        try:
            
            # Construct the Lighthouse CLI command
            command = [
                'lighthouse',
                url,
                '--quiet' if quiet else '',
                '--chrome-flags="--headless --no-sandbox"',
                '--preset=perf',
                '--emulated-form-factor={0}'.format(form_factor),
                '--output=json',
                '--output=html',
                '--output-path={0}'.format(self.report_path),
            ]

            # Run the Lighthouse CLI command
            command = command + additional_settings
            subprocess.check_call(' '.join(command), shell=True)
        except subprocess.CalledProcessError as exc:
            # If an error occurs, raise a RuntimeError with a detailed error message
            msg = f"Command '{exc.cmd}' returned an error code: {exc.returncode}, output: {exc.output}"
            raise RuntimeError(msg)

    def _get_report(self) -> Union[LighthouseReport, None]:
        """
        Copy the generated report files to the output directory (if specified) and return a LighthouseReport object 
        containing the report data.

        Returns:
            LighthouseReport or None: A LighthouseReport object containing the report data, or None if the report 
            could not be loaded.
        
        Raises:
            Exception: If there is an error loading the report data from the JSON file.
        """
        
        # Copy the HTML report to the output directory if output_type is 'html' or 'both'
        if self.output_type in ["html","both"]:
            os.makedirs(self.output_dir,exist_ok=True)
            shutil.copy(self.report_path+".report.html", os.path.join(self.output_dir,self.report_name+".html"))
            
        # Copy the JSON report to the output directory if output_type is 'json' or 'both'
        if self.output_type in ["json","both"]:
            os.makedirs(self.output_dir,exist_ok=True)
            shutil.copy(self.report_path+".report.json", os.path.join(self.output_dir,self.report_name+".json"))
            
        # Load the report data from the JSON file and return a LighthouseReport object
        try:
            with open(self.report_path+".report.json", 'r') as fil:
                data = json.load(fil)
                return LighthouseReport(data,self.timings)
        except Exception as e:
            print(f"Error loading report: {e}")
            return None

    def _clean(self) -> None:
        """
        Clean up any temporary files created during the performance test.
        """
        
        # Remove the JSON and HTML report files generated by Lighthouse
        os.remove(self.report_path+".report.json")
        os.remove(self.report_path+".report.html")

class LighthouseRepeatRunner:
    """
    A class for running performance tests using Lighthouse, multiple times and returning 
    an averaged report.

    Attributes:
        report (namedtuple): A named tuple containing the averaged timings and score of the 
            Lighthouse performance test. The named tuple has two fields: 'timings' and 'score'.
            Both fields are dictionaries that map performance metric names to their average 
            value across all test repeats.
    """
    
    def __init__(self, url: str, form_factor: str = 'mobile', quiet: bool = True,
                 additional_settings: List[str] = [], repeats: int = 3, timings:List[str]=BASE_TIMINGS):
        """ 
        Initialize a new LighthouseRepeatRunner object.
        
        Args:
            url (str): The URL to test.
            
            form_factor (str, optional): The form factor to use for the test. Either 'mobile' or 'desktop'.
                Defaults to 'mobile'.
                
            quiet (bool, optional): Whether to suppress output to stdout. Defaults to True.
            
            additional_settings (list of str, optional): A list of additional parameters to use for the test.
                Defaults to an empty list.
                
            repeats (int, optional): The number of times to repeat the Lighthouse performance test. 
                Defaults to 3.
            
            timings (list, optional): A list of performance metrics to measure during the test.
                Defaults to BASE_TIMINGS, which includes several common metrics.
        """
        
        # Initialize an empty list to store the Lighthouse reports from each test repeat
        reports = []
        
        # Use tqdm to display a progress bar while running the test repeats
        progress = tqdm(range(1, repeats + 1), desc='Repeating test')

        # Run the Lighthouse test the specified number of times and store the reports in a list
        for i in progress:
            progress.set_description('Run {0}/{1}'.format(i, repeats))
            reports.append(LighthouseRunner(url, form_factor=form_factor,
                                            quiet=quiet,
                                            additional_settings=additional_settings,timings=timings).report) 

        # Calculate the average values for each performance metric across all test repeats
        report = namedtuple('LighthouseAveragedReport', 'timings, score')
        self.report = report(
            timings=self._get_average([getattr(x, 'timings') for x in reports]),
            score=self._get_average([getattr(x, 'score') for x in reports])
        )

    def _get_average(self, obj_lst: List[Dict[str, Any]]) -> Dict[str, Any]:
        """
        Computes the average value for each performance metric across all Lighthouse reports.
        
        Args:
            obj_lst (list of dict): A list of dictionaries, where each dictionary contains performance
                metrics for a single Lighthouse report.
        
        Returns:
            dict: A dictionary containing the average value for each performance metric across all reports. 
                The keys of the dictionary are the names of the performance metrics, and the values are the 
                average value of that metric across all reports.
        """
        
        ret = {}
        for key in obj_lst[0].keys():
            lst = [x.get(key) for x in obj_lst]
            ret[key] = reduce(lambda a, b: a + b, lst) / len(lst)
        return ret

class BatchRunner:
    """
    Runs Lighthouse reports for a batch of URLs and form factors.
    """
    
    def __init__(self, urls: List[str], form_factors: List[str], quiet: bool = True,
                 additional_settings: dict = None, repeats: int = 3, timings:List[str]=BASE_TIMINGS):
        """
        Args:
            urls (list(str)): List of URLs to run Lighthouse reports on.
            
            form_factors (list(str)): List of form factors to run Lighthouse reports on.
            
            quiet (bool, optional): Whether to suppress Lighthouse output. Defaults to True.
            
            additional_settings (dict, optional): Additional Lighthouse settings to use for all reports.
                Defaults to None.
                
            repeats (int, optional): Number of times to repeat each report. Defaults to 3.
            
            timings (list, optional): A list of performance metrics to measure during the test.
                Defaults to BASE_TIMINGS, which includes several common metrics.
        """
        
        # Ensure urls and form_factors are lists
        if not isinstance(form_factors, (list, )):
            form_factors = list(form_factors)
        if not isinstance(urls, (list, )):
            urls = list(urls)

        # Iterate over all combinations of urls and form_factors
        all_combinations = tqdm(list(product(urls, form_factors)),
                                desc='Running urls')

        self.reports = []

        for url, factor in all_combinations:
            all_combinations.set_description('{0} | {1}'.format(url, factor))
            
            # Run the Lighthouse report for this combination and append it to reports list
            report = LighthouseRepeatRunner(url, factor, quiet,
                                            additional_settings,
                                            repeats=repeats,timings=timings).report
            self.reports.append((url, factor, report))


demo/demo_runner.py
from lighthouse import LighthouseRunner


url="https://github.com"

TIMINGS = [
    'first-contentful-paint',
    'speed-index',
    'interactive',
]

# settings are as the same as the pagespeed.web.dev
additional_settings = [
"--throttling.cpuSlowdownMultiplier=4",
"--throttling.connectionType=4g",
"--throttling.downloadThroughputKbps=1638.4",
"--throttling.rttMs=150",
"--screenEmulation.disabled=false",
"--screenEmulation.width=412",
"--screenEmulation.height=823",
"--screenEmulation.deviceScaleFactor=1.75",
"--axe.enable=true",
"--axe.version=4.7.0",
]

runner = LighthouseRunner(url,"mobile",quiet=False,additional_settings=additional_settings,
                          timings=TIMINGS,output_type='both',output_dir='./outputs')

print(runner.report.score)
print(runner.report.timings)
print(runner.report.audits(0.86)['performance'].passed)
